app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,109 @@
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppStartupSourceAndType]and finally report them on Activity's onResume() function.
+ * If application wasn't backgrounded before onResume() is called, we don't report the metric.
+ */
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var appOpenedAllSourceStartup = AppStartupSourceAndType()
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
+        setAppStartupType()
+        setStartupSourceFromIntent(safeIntent, false)
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
+        setAppStartupType()
+        setStartupSourceFromIntent(safeIntent, true)
+    }
+
+    fun onHomeActivityOnRestart() {
+        appOpenedAllSourceStartup.type = Type.HOT
+        appOpenedAllSourceStartup.hasSavedInstance = false
+    }
+
+    fun receivedIntentInHomeActivity(safeIntent: SafeIntent) {
+        setStartupSourceFromIntent(safeIntent, false)
+    }
+
+    private fun setAppStartupType() {
+        appOpenedAllSourceStartup.type = if (wasAppCreateCalledBeforeActivityCreate) COLD else WARM
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    private fun setStartupSourceFromIntent(
+        intent: SafeIntent,
+        isExternalAppBrowserActivity: Boolean
+    ) {
+        appOpenedAllSourceStartup.source = when {
+            isExternalAppBrowserActivity -> Source.CUSTOM_TAB
+            intent.isLauncherIntent -> Source.APP_ICON
+            intent.action == Intent.ACTION_VIEW -> Source.LINK
+            else -> Source.UNKNOWN
+        }
+    }
+
+    /**
+     * The reason we record metric on resume is because we need to wait for onNewIntent(), and
+     * we are not guaranteed that onNewIntent() will be called before or after onStart().
+     * However we are guaranteed onResume() will be called after onNewIntent(). Source:
+     * https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)
+     *
+     */
+    fun onHomeActivityOnResume() {
+        recordMetric()
+    }
+
+    private fun recordMetric() {
+        if (!isMetricRecordedSinceAppWasForegrounded) {
+            metrics.track(appOpenedAllSourceStartup)
+            isMetricRecordedSinceAppWasForegrounded = true
+        }
+        // we don't want any weird previous states to persist on our next metric record.
```

Comment:
Good idea – you can invalidate some of what I said above but I'm not going to rewrite it because I already typed too much 😅

note: If we switch to an immutable data container, we'd probably assign to the "no data" state here instead.